zebra-chain/src/serialization/arbitrary.rs

```diff
@@ -72,3 +74,40 @@ impl Arbitrary for CompactSizeMessage {
 
     type Strategy = BoxedStrategy<Self>;
 }
+
+/// Allocate a maximum-sized vector of cloned `item`s, and serialize that array.
+///
+/// Returns the following calculations on `item`:
+///   smallest_disallowed_vec_len
+///   smallest_disallowed_serialized_len
+///   largest_allowed_vec_len
+///   largest_allowed_serialized_len
+///
+/// For varible-size types, `largest_allowed_serialized_len` might not fit within
+/// `MAX_PROTOCOL_MESSAGE_LEN` or `MAX_BLOCK_SIZE`.
+pub fn max_allocation_is_big_enough<T>(item: T) -> (usize, usize, usize, usize)
+where
+    T: TrustedPreallocate + ZcashSerialize + Clone,
+{
+    let max_allocation: usize = T::max_allocation().try_into().unwrap();
+    let mut smallest_disallowed_vec = vec![item; max_allocation + 1];
+
+    let smallest_disallowed_serialized = smallest_disallowed_vec
+        .zcash_serialize_to_vec()
+        .expect("Serialization to vec must succeed");
+    let smallest_disallowed_vec_len = smallest_disallowed_vec.len();
+
+    // Create largest_allowed_vec by removing one element from smallest_disallowed_vec without copying (for efficiency)
+    smallest_disallowed_vec.pop();
+    let largest_allowed_vec = smallest_disallowed_vec;
+    let largest_allowed_serialized = largest_allowed_vec
+        .zcash_serialize_to_vec()
+        .expect("Serialization to vec must succeed");
+
+    (
+        smallest_disallowed_vec_len,
+        smallest_disallowed_serialized.len(),
+        largest_allowed_vec.len(),
+        largest_allowed_serialized.len(),
+    )
+}
```

Comment:
Suggestion: would it help to collect the assertions into helper functions as well?

```suggestion
}

pub fn assert_max_allocation_is_big_enough<T, const LIMIT: usize>(item: T)
where
    T: TrustedPreallocate + ZcashSerialize + Clone,
{
    let (
        smallest_disallowed_vec_len,
        smallest_disallowed_serialized_len,
        largest_allowed_vec_len,
        largest_allowed_serialized_len,
    ) = max_allocation_is_big_enough(item);

    // Check that our smallest_disallowed_vec is only one item larger than the limit
    assert_eq!(((smallest_disallowed_vec_len - 1) as u64), T::max_allocation());
    // Check that our smallest_disallowed_vec is too big to send as a protocol message
    assert!(smallest_disallowed_serialized_len > LIMIT);

    // Check that our largest_allowed_vec contains the maximum number of items
    assert_eq!((largest_allowed_vec_len as u64), T::max_allocation());
    // Check that our largest_allowed_vec is small enough to send as a protocol message
    assert!(largest_allowed_serialized_len <= LIMIT);
}

pub fn prop_assert_max_allocation_is_big_enough<T, const LIMIT: usize>(
    item: T,
) -> Result<(), TestCaseError>
where
    T: TrustedPreallocate + ZcashSerialize + Clone,
{
    let (
        smallest_disallowed_vec_len,
        smallest_disallowed_serialized_len,
        largest_allowed_vec_len,
        largest_allowed_serialized_len,
    ) = max_allocation_is_big_enough(item);

    // Check that our smallest_disallowed_vec is only one item larger than the limit
    prop_assert_eq!(((smallest_disallowed_vec_len - 1) as u64), T::max_allocation());
    // Check that our smallest_disallowed_vec is too big to send as a protocol message
    prop_assert!(smallest_disallowed_serialized_len > LIMIT);

    // Check that our largest_allowed_vec contains the maximum number of items
    prop_assert_eq!((largest_allowed_vec_len as u64), T::max_allocation());
    // Check that our largest_allowed_vec is small enough to send as a protocol message
    prop_assert!(largest_allowed_serialized_len <= LIMIT);
    
    Ok(())
}
```

Reply:
Not really, because the line numbers wouldn't correspond to each test.
Also sometimes the limits are different or skipped.

Reply:
I think line numbers shouldn't be an issue, as long as we only call this once per test. We can figure out what went wrong with the test name and the helper function's line number.

Regarding skipped assertions, that shouldn't be an issue, right? I mean, there's no case that having the extra assertion would make the test fail, I think. And regarding different limits (with the exception of `u8`), shouldn't it just be a matter of using the `LIMIT` constant?

But don't let this block you. I'll try to finish the rest of the review ASAP :+1:

Reply:
> Regarding skipped assertions, that shouldn't be an issue, right? I mean, there's no case that having the extra assertion would make the test fail, I think.

Preallocation uses the maximum length allowed for an array of the type, based on the minimum size of the type. So variable-length types disable the final assertion, because some instances of the type can exceed the limit. (But some don't.)

But in any case, this change is out of scope, because `addrv2` serialization needs to be finished this sprint. And this PR is meant to just be a trivial refactor to make `addrv2` tests easier to write.